File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/impl/StoredLoader.java
package ru.kmorozov.library.data.loader.impl;

import org.springframework.beans.factory.annotation.Autowired;
import ru.kmorozov.gbd.logger.Logger;
import ru.kmorozov.library.data.loader.netty.EventSender;
import ru.kmorozov.library.data.model.book.Book;
import ru.kmorozov.library.data.model.book.BookInfo;
import ru.kmorozov.library.data.model.book.BookInfo.BookFormat;
import ru.kmorozov.library.data.model.book.Category;
import ru.kmorozov.library.data.model.book.Storage;
import ru.kmorozov.library.data.model.book.StorageInfo;
import ru.kmorozov.library.data.repository.BooksRepository;
import ru.kmorozov.library.data.repository.CategoryRepository;
import ru.kmorozov.library.data.repository.StorageRepository;
import ru.kmorozov.library.utils.BookUtils;

import java.io.IOException;
import java.util.Date;
import java.util.stream.Stream;

public abstract class StoredLoader extends BaseLoader{

    private static final Logger logger = Logger.getLogger(StoredLoader.class);

    @Autowired
    protected CategoryRepository categoryRepository;

    @Autowired
    protected StorageRepository storageRepository;

    @Autowired
    protected BooksRepository booksRepository;

    public void clear() {
        long categoryCount = this.categoryRepository.count();
        long storageCount = this.storageRepository.count();
        long booksCount = this.booksRepository.count();

        if (0L < categoryCount) {
            StoredLoader.logger.info("Categories loaded: " + categoryCount);
            this.categoryRepository.deleteAll();
        }

        if (0L < storageCount) {
            StoredLoader.logger.info("Storages loaded: " + storageCount);
            this.storageRepository.deleteAll();
        }

        if (0L < booksCount) {
            StoredLoader.logger.info("Books loaded: " + storageCount);
            this.booksRepository.deleteAll();
        }
    }

    private Category getOrCreateCategory(String name) {
        Category category = this.categoryRepository.findOneByName(name);
        if (null == category) {
            category = new Category();
            category.setName(name);

            this.categoryRepository.save(category);
        }

        return category;
    }

    private Storage getOrCreateStorage(ServerItem serverItem) {
        Storage storage = this.storageRepository.findByUrl(serverItem.getUrl());
        return StoredLoader.fillStorage(null == storage ? new Storage() : storage, serverItem);
    }

    protected static Storage fillStorage(Storage storage, ServerItem serverItem) {
        storage.setStorageType(serverItem.getStorageType());
        storage.setUrl(serverItem.getUrl());
        storage.setName(serverItem.getName());
        storage.setLastModifiedDateTime(serverItem.getLastModifiedDateTime());
        storage.setStorageInfo(new StorageInfo(serverItem.getFilesCount()));
        storage.getStorageInfo().setLastChecked(System.currentTimeMillis());

        return storage;
    }

    protected Category getCategoryByServerItem(ServerItem serverItem) {
        Category category = this.getOrCreateCategory(serverItem.getName());
        Storage storage = this.getOrCreateStorage(serverItem);

        storage.addCategory(category);

        this.storageRepository.save(storage);

        Storage parentStorage = null == serverItem.getParent() ? null : this.storageRepository.findByUrl(serverItem.getParent().getUrl());

        if (null != parentStorage) {
            storage.setParent(parentStorage);
            this.storageRepository.save(storage);

            category.addParents(parentStorage.getCategories());
            this.categoryRepository.save(category);
        }

        category.addStorage(storage);
        this.categoryRepository.save(category);

        return storage.getMainCategory();
    }

    protected abstract Stream<ServerItem> getItemsStreamByStorage(Storage storage) throws IOException;

    public abstract void processLinks();

    public abstract boolean postponedLinksLoad();

    protected void updateStorage(Storage storage) throws IOException {
        StorageInfo storageInfo = null == storage.getStorageInfo() ? new StorageInfo() : storage.getStorageInfo();

        this.getItemsStreamByStorage(storage)
                .filter(ServerItem::isLoadableOrLink)
                .forEach(serverItem -> {
                    if (!serverItem.isDirectory()) {
                        BookFormat bookFormat = BookUtils.getFormat(serverItem.getName());
                        if (BookFormat.UNKNOWN != bookFormat) {
                            Book existBook = this.booksRepository.findOneByBookInfoPath(serverItem.getUrl());
                            if (null == existBook) {
                                Book book = new Book();

                                BookInfo bookInfo = new BookInfo();
                                bookInfo.setFileName(serverItem.getName());
                                bookInfo.setPath(serverItem.getUrl());
                                bookInfo.setFormat(bookFormat);
                                bookInfo.setLastModifiedDateTime(serverItem.getLastModifiedDateTime());
                                bookInfo.setSize(serverItem.getSize());

                                book.setBookInfo(bookInfo);
                                book.setStorage(storage);

                                if (book.isLink() && !this.postponedLinksLoad()) {
                                    this.links.add(serverItem.getOriginalItem());
                                    EventSender.INSTANCE.sendInfo(StoredLoader.logger, "Added link " + serverItem.getName());
                                } else {
                                    this.booksRepository.save(book);
                                    storage.getStorageInfo().incFilesCount();
                                    EventSender.INSTANCE.sendInfo(StoredLoader.logger, "Added file " + serverItem.getName());
                                }
                            } else {
                                Date oldDate = existBook.getBookInfo().getLastModifiedDateTime();
                                Date newDate = serverItem.getLastModifiedDateTime();
                                boolean dateCondition = null == oldDate || oldDate.before(newDate);

                                long oldSize = existBook.getBookInfo().getSize();
                                long newSize = serverItem.getSize();
                                boolean sizeCondition = 0L == oldSize || oldSize != newSize;

                                final boolean storageCondition = !existBook.getStorage().equals(storage);
                                final boolean nameCondition = serverItem.getName().equals(existBook.getBookInfo().getFileName());

                                if (dateCondition || sizeCondition || storageCondition || nameCondition) {
                                    existBook.getBookInfo().setFileName(serverItem.getName());
                                    existBook.getBookInfo().setLastModifiedDateTime(newDate);
                                    existBook.getBookInfo().setSize(newSize);
                                    existBook.setStorage(storage);
                                    this.booksRepository.save(existBook);

                                    EventSender.INSTANCE.sendInfo(StoredLoader.logger, "Updated file " + serverItem.getName());
                                }
                            }
                        }
                    }
                });

        storageInfo.setLastChecked(System.currentTimeMillis());

        storage.setStorageInfo(storageInfo);
        this.storageRepository.save(storage);
    }
}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/impl/StoredLoader.kt
package ru.kmorozov.library.data.loader.impl

import org.springframework.beans.factory.annotation.Autowired
import ru.kmorozov.gbd.logger.Logger
import ru.kmorozov.library.data.model.book.*
import ru.kmorozov.library.data.repository.BooksRepository
import ru.kmorozov.library.data.repository.CategoryRepository
import ru.kmorozov.library.data.repository.StorageRepository
import ru.kmorozov.library.utils.BookUtils
import java.io.IOException
import java.util.function.Predicate
import java.util.stream.Stream

abstract class StoredLoader : BaseLoader() {

    @Autowired
    protected lateinit var categoryRepository: CategoryRepository

    @Autowired
    protected lateinit var storageRepository: StorageRepository

    @Autowired
    protected lateinit var booksRepository: BooksRepository

    fun clear() {
        val categoryCount = categoryRepository.count()
        val storageCount = storageRepository.count()
        val booksCount = booksRepository.count()

        if (0L < categoryCount) {
            logger.info("Categories loaded: $categoryCount")
            categoryRepository.deleteAll()
        }

        if (0L < storageCount) {
            logger.info("Storages loaded: $storageCount")
            storageRepository.deleteAll()
        }

        if (0L < booksCount) {
            logger.info("Books loaded: $storageCount")
            booksRepository.deleteAll()
        }
    }

    private fun getOrCreateCategory(name: String): Category {
        var category: Category? = categoryRepository.findOneByName(name)
        if (null == category) {
            category = Category()
            category.name = name

            categoryRepository.save(category)
        }

        return category
    }

    private fun getOrCreateStorage(serverItem: ServerItem): Storage {
        val storage = storageRepository.findByUrl(serverItem.url)
        return fillStorage(storage ?: Storage(), serverItem)
    }

    protected fun getCategoryByServerItem(serverItem: ServerItem): Category? {
        val category = getOrCreateCategory(serverItem.name)
        val storage = getOrCreateStorage(serverItem)

        storage.addCategory(category)

        storageRepository.save(storage)

        val parentStorage = if (null == serverItem.parent) null else storageRepository.findByUrl(serverItem.parent!!.url)

        if (null != parentStorage) {
            storage.parent = parentStorage
            storageRepository.save(storage)

            category.addParents(parentStorage.getCategories()!!)
            categoryRepository.save(category)
        }

        category.addStorage(storage)
        categoryRepository.save(category)

        return storage.mainCategory
    }

    @Throws(IOException::class)
    protected abstract fun getItemsStreamByStorage(storage: Storage): Stream<ServerItem>

    abstract fun processLinks()

    abstract fun postponedLinksLoad(): Boolean

    @Throws(IOException::class)
    protected fun updateStorage(storage: Storage) {
        val storageInfo = if (null == storage.storageInfo) StorageInfo() else storage.storageInfo

        getItemsStreamByStorage(storage)
                .filter(Predicate<ServerItem> { it.isLoadableOrLink })
                .forEach { serverItem ->
                    if (!serverItem.isDirectory) {
                        val bookFormat = BookUtils.getFormat(serverItem.name)
                        if (BookInfo.BookFormat.UNKNOWN !== bookFormat) {
                            val existBook = booksRepository.findOneByBookInfoPath(serverItem.url)
                            if (null == existBook) {
                                val book = Book()

                                val bookInfo = BookInfo()
                                bookInfo.fileName = serverItem.name
                                bookInfo.path = serverItem.url
                                bookInfo.format = bookFormat
                                bookInfo.lastModifiedDateTime = serverItem.lastModifiedDateTime
                                bookInfo.size = serverItem.size

                                book.bookInfo = bookInfo
                                book.storage = storage

                                if (book.isLink && !postponedLinksLoad()) {
                                    links.add(serverItem.originalItem)
                                } else {
                                    booksRepository.save(book)
                                    storage.storageInfo!!.incFilesCount()
                                }
                            } else {
                                val oldDate = existBook.bookInfo.lastModifiedDateTime
                                val newDate = serverItem.lastModifiedDateTime
                                val dateCondition = null == oldDate || oldDate.before(newDate)

                                val oldSize = existBook.bookInfo.size
                                val newSize = serverItem.size
                                val sizeCondition = 0L == oldSize || oldSize != newSize

                                val storageCondition = existBook.storage != storage
                                val nameCondition = serverItem.name == existBook.bookInfo.fileName

                                if (dateCondition || sizeCondition || storageCondition || nameCondition) {
                                    existBook.bookInfo.fileName = serverItem.name
                                    existBook.bookInfo.lastModifiedDateTime = newDate
                                    existBook.bookInfo.size = newSize
                                    existBook.storage = storage
                                    booksRepository.save(existBook)
                                }
                            }
                        }
                    }
                }

        storageInfo!!.lastChecked = System.currentTimeMillis()

        storage.storageInfo = storageInfo
        storageRepository.save(storage)
    }

    companion object {

        private val logger = Logger.getLogger(StoredLoader::class.java)

    }

    protected fun fillStorage(storage: Storage, serverItem: ServerItem): Storage {
        storage.storageType = serverItem.storageType
        storage.url = serverItem.url
        storage.name = serverItem.name
        storage.lastModifiedDateTime = serverItem.lastModifiedDateTime
        storage.storageInfo = StorageInfo(serverItem.filesCount)
        storage.storageInfo!!.lastChecked = System.currentTimeMillis()

        return storage
    }
}
